### PR TITLE
Support multipart parsing for RequestBodyParserMiddleware (file uploads)

### LINKED ISSUE #105: Parse request body
We should provide an *easier* way to access the parsed request body, i.e. `$_POST` values from HTML forms, file uploads etc.

This depends on detecting actual message lengths first, see #104.

I'm filing this mostly for the reference, there's been some discussion to wait for PSR-7 message support (#28) first.

### PR DESCRIPTION
**Pull request by WyriHaximus**
Follow up on #220 adding multipart support for forms and file uploads
Supersedes / Closes #41 
Implements / Closes #105 
Implements / Closes #117 

Todo: 
- [X] Update middleware adding multipart parsing
- [x] Readme

### CODE REVIEW

**Comment by WyriHaximus** [comment]
@jsor @clue added readme changes for file uploads

File: src/Middleware/MultipartParser.php [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+
+namespace React\Http\Middleware;
+
+use Psr\Http\Message\RequestInterface;
+use Psr\Http\Message\ServerRequestInterface;
+use React\Http\HttpBodyStream;
+use React\Http\UploadedFile;
+use RingCentral\Psr7;
+
+final class MultipartParser
```
Comment (clue) [inline src/Middleware/MultipartParser.php]: `@internal`?

File: README.md [hunk 2]
```diff
@@ -746,6 +746,27 @@ $parsedBody = [
...
+    'file' => new UploadedFile(/**...**/),
+    'first' => [
+        'second' => new UploadedFile(/**...**/),
+        'third' => [
+            new UploadedFile(/**...**/),
+            new UploadedFile(/**...**/),
+        ],
+    ],
+];
+```
```
Comment (clue) [inline README.md]: The documentation could probably use a simpler example :-)

File: src/Middleware/MultipartParser.php [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+
+    /**
+     * @var string
+     */
+    protected $ending = '';
+
+    /**
+     * @var int
+     */
+    protected $endingSize = 0;
```
Comment (clue) [inline src/Middleware/MultipartParser.php]: leftover?

File: src/Middleware/MultipartParser.php [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+{
+    /**
+     * @var string
+     */
+    protected $buffer = '';
+
+    /**
+     * @var string
+     */
+    protected $ending = '';
```
Comment (clue) [inline src/Middleware/MultipartParser.php]: leftover?

File: src/Middleware/MultipartParser.php [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+     */
+    protected $endingSize = 0;
+
+    /**
+     * @var string
+     */
+    protected $boundary;
+
+    /**
+     * @var RequestInterface
```
Comment (clue) [inline src/Middleware/MultipartParser.php]: `ServerRequestInterface`?

File: src/Middleware/MultipartParser.php [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+        $this->determineStartMethod();
+
+        return $this->request;
+    }
+
+    private function determineStartMethod()
+    {
+        if (!$this->request->hasHeader('content-type')) {
+            $this->findBoundary();
+        }
```
Comment (clue) [inline src/Middleware/MultipartParser.php]: leftover?

File: src/Middleware/MultipartParser.php [hunk 7]
```diff
@@ -0,0 +1,266 @@
...
+            $boundary = substr($this->buffer, 2, strpos($this->buffer, "\r\n"));
+            $boundary = substr($boundary, 0, -2);
+            $this->setBoundary($boundary);
+            $this->parseBuffer();
+        }
+    }
+
+    private function parseBuffer()
+    {
+        $chunks = preg_split('/-+' . $this->boundary . '/', $this->buffer);
```
Comment (clue) [inline src/Middleware/MultipartParser.php]: The boundary may potentially contain characters that would be interpreted as part of the regex here

File: src/Middleware/MultipartParser.php [hunk 8]
```diff
@@ -0,0 +1,266 @@
...
+            $chunk = $this->stripTrailingEOL($chunk);
+            $this->parseChunk($chunk);
+        }
+    }
+
+    private function parseChunk($chunk)
+    {
+        if ($chunk === '') {
+            return;
+        }
```
Comment (clue) [inline src/Middleware/MultipartParser.php]: Plenty of these checks, use is unclear and also looks untested?

File: src/Middleware/RequestBodyParserMiddleware.php [hunk 9]
```diff
@@ -9,11 +9,20 @@
...
             return $next($this->parseFormUrlencoded($request));
         }
 
+        if ($type === 'multipart/form-data') {
+            return $next(MultipartParser::parseRequest($request));
+        }
+
+        if ($type === 'multipart/mixed') {
+            return $next(MultipartParser::parseRequest($request));
+        }
```
Comment (clue) [inline src/Middleware/RequestBodyParserMiddleware.php]: Couldn't find a *relevant* mention of `multipart/mixed` in https://tools.ietf.org/html/rfc7578#section-4.3 and https://www.w3.org/TR/html5/forms.html#multipart-form-data, is this needed?

Comment (WyriHaximus) [reply src/Middleware/RequestBodyParserMiddleware.php]: IIRC I've seen both during development, I'll run some tests when I get home to be sure of this 👍

File: src/Middleware/MultipartParser.php (hunk 8 above)
Comment (WyriHaximus) [reply src/Middleware/MultipartParser.php]: Removing that instantly breaks the tests

File: src/Middleware/MultipartParser.php (hunk 7 above)
Comment (WyriHaximus) [reply src/Middleware/MultipartParser.php]: Good catch, I'll fortify it 👍

Comment (WyriHaximus) [reply src/Middleware/MultipartParser.php]: Actually I've swapped it out with an explode for simplicity

**Comment by WyriHaximus** [comment]
Ping @clue, all your concens have been address and where needed code has been updated.

**Comment by andig** [comment]
ping @clue would it be possible to proceed with this PR?

**Comment by WyriHaximus** [comment]
🎉

**Pull request merged**